application/controllers/ResponsesController.php

```diff
@@ -192,7 +194,21 @@ public function actionView(int $surveyId, int $id, string $browseLang = ''): voi
                     'code' => viewHelper::getFieldCode($field, ['LEMcompat' => true])
                 ];
             } elseif ($field['aid'] !== 'filecount') {
-                $qidattributes = QuestionAttribute::model()->getQuestionAttributes($field['qid']);
+                $qids[] = $field['qid'];
+                $fields[] = $field;
+            } else {
+                $fnames[] = [$field['fieldname'], gT("File count")];
```

Comment:
It's another fix here ? What happen for comments and other ?

Maybe remove this fix (if it's not needed) ?

Reply:
@Shnoulle 

> It's another fix here ? What happen for comments and other ?
> 
> Maybe remove this fix (if it's not needed) ?

We had this:

```
        $fnames[] = ["completed", gT("Completed"), "0"];

        foreach ($fieldmap as $field) {
            if ($field['fieldname'] == 'lastpage' || $field['fieldname'] == 'submitdate') {
                continue;
            }
            if ($field['type'] == 'interview_time') {
                continue;
            }
            if ($field['type'] == 'page_time') {
                continue;
            }
            if ($field['type'] == 'answer_time') {
                continue;
            }

            //$question = $field['question'];
            $question = viewHelper::getFieldText($field);

            if ($field['type'] != Question::QT_VERTICAL_FILE_UPLOAD) {
                $fnames[] = [
                    $field['fieldname'],
                    viewHelper::getFieldText($field),
                    'code' => viewHelper::getFieldCode($field, ['LEMcompat' => true])
                ];
            } elseif ($field['aid'] !== 'filecount') {
                $qidattributes = QuestionAttribute::model()->getQuestionAttributes($field['qid']);

                for ($i = 0; $i < $qidattributes['max_num_of_files']; $i++) {
                    $filenum = sprintf(gT("File %s"), $i + 1);
                    if ($qidattributes['show_title'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Title') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(title)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "title",
                            "index"    => $i
                        ];
                    }

                    if ($qidattributes['show_comment'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Comment') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(comment)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "comment",
                            "index"    => $i
                        ];
                    }

                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File name') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(name)',
                        "type"     => "|",
                        "metadata" => "name",
                        "index"    => $i,
                        'qid'      => $field['qid']
                    ];
                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File size') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(size)',
                        "type"     => "|",
                        "metadata" => "size",
                        "index"    => $i
                    ];
                }
            } else {
                $fnames[] = [$field['fieldname'], gT("File count")];
            }
        }
```

and now it is:

```
        $fnames[] = ["completed", gT("Completed"), "0"];
        $qids = [];
        $fields = [];

        foreach ($fieldmap as $field) {
            if ($field['fieldname'] == 'lastpage' || $field['fieldname'] == 'submitdate') {
                continue;
            }
            if ($field['type'] == 'interview_time') {
                continue;
            }
            if ($field['type'] == 'page_time') {
                continue;
            }
            if ($field['type'] == 'answer_time') {
                continue;
            }

            //$question = $field['question'];
            $question = viewHelper::getFieldText($field);

            if ($field['type'] != Question::QT_VERTICAL_FILE_UPLOAD) {
                $fnames[] = [
                    $field['fieldname'],
                    viewHelper::getFieldText($field),
                    'code' => viewHelper::getFieldCode($field, ['LEMcompat' => true])
                ];
            } elseif ($field['aid'] !== 'filecount') {
                $qids[] = $field['qid'];
                $fields[] = $field;
            } else {
                $fnames[] = [$field['fieldname'], gT("File count")];
            }
        }

        if (count($qids)) {
            $rawQuestions = Question::model()->findAllByPk($qids);
            $questions = [];
            foreach ($rawQuestions as $rawQuestion) {
                $questions[$rawQuestion->qid] = $rawQuestion;
            }
            foreach ($fields as $field) {
                $qidattributes = QuestionAttribute::model()->getQuestionAttributes($questions[$field['qid']]);

                for ($i = 0; $i < $qidattributes['max_num_of_files']; $i++) {
                    $filenum = sprintf(gT("File %s"), $i + 1);
                    if ($qidattributes['show_title'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Title') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(title)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "title",
                            "index"    => $i
                        ];
                    }

                    if ($qidattributes['show_comment'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Comment') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(comment)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "comment",
                            "index"    => $i
                        ];
                    }

                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File name') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(name)',
                        "type"     => "|",
                        "metadata" => "name",
                        "index"    => $i,
                        'qid'      => $field['qid']
                    ];
                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File size') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(size)',
                        "type"     => "|",
                        "metadata" => "size",
                        "index"    => $i
                    ];
                }
            }
        }
```

So we had a loop of fieldmap and for each map we had a structure of

```
            if ($field['type'] != Question::QT_VERTICAL_FILE_UPLOAD) {
                ...
            } elseif ($field['aid'] !== 'filecount') {
                ...
            } else {
                ...
            }
```

This structure is left unchanged, but in the else-if branch we store the qid and the field in appropriate arrays, so instead later we get the questions together and pass the question object to `getQuestionAttributes` instead of passing the qid value.

So after the loop where we planned what questions are to be changed, we check whether there was at least one qid to handle. If so, we load the questions together, map them to their ids to preserve the order and then loop the fields array we created for this purpose and proceed like we did before with:

```
                $qidattributes = QuestionAttribute::model()->getQuestionAttributes($questions[$field['qid']]);

                for ($i = 0; $i < $qidattributes['max_num_of_files']; $i++) {
                    $filenum = sprintf(gT("File %s"), $i + 1);
                    if ($qidattributes['show_title'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Title') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(title)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "title",
                            "index"    => $i
                        ];
                    }

                    if ($qidattributes['show_comment'] == 1) {
                        $fnames[] = [
                            $field['fieldname'],
                            "{$filenum} - {$question} (" . gT('Comment') . ")",
                            'code'     => viewHelper::getFieldCode($field) . '(comment)',
                            "type"     => Question::QT_VERTICAL_FILE_UPLOAD,
                            "metadata" => "comment",
                            "index"    => $i
                        ];
                    }

                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File name') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(name)',
                        "type"     => "|",
                        "metadata" => "name",
                        "index"    => $i,
                        'qid'      => $field['qid']
                    ];
                    $fnames[] = [
                        $field['fieldname'],
                        "{$filenum} - {$question} (" . gT('File size') . ")",
                        'code'     => viewHelper::getFieldCode($field) . '(size)',
                        "type"     => "|",
                        "metadata" => "size",
                        "index"    => $i
                    ];
                }
            }
```

the point being that we pass the question object to `getQuestionAttributes` instead of the id, so we will not individually load the question objects from the database, but we load them together.

Reply:
Sorry, previous line is under the added part :)

![Capture du 2024-02-29 12-29-02](https://github.com/LimeSurvey/LimeSurvey/assets/1439428/5d1b5808-e343-4e11-b92e-b0f494741ded)